app/assets/javascripts/helpers/profile_photo_uploader.es6

```diff
@@ -0,0 +1,250 @@
+      // set the preferred size style of the cropper based on picture orientation
+      this.picture.setAttribute("style", (this.picture.naturalHeight > this.picture.naturalWidth ?
+        "max-height:600px;max-width:none;" : "max-width:600px;"));
+      this.buildControls();
+
+      this.setLoading(false);
+
+      // eslint-disable-next-line no-undef
+      this.cropper = new Cropper(this.picture, {
+        aspectRatio: 1,
+        zoomable: false,
+        autoCropArea: 1,
+        preview: ".previews > div"
+      });
+    };
+    this.picture.setAttribute("src", imageData);
+  }
+
+  /**
+   * Creates image manipulation controls and previews.
+   */
+  buildControls() {
+    this.controls = {
+      rotateLeft: this.createButton("ccw", () => this.cropper.rotate(-45)),
+      rotateRight: this.createButton("cw", () => this.cropper.rotate(45)),
+      reset: this.createButton("cycle", () => this.cropper.reset()),
+      accept: this.createButton("check", () => this.cropImage()),
+      cancel: this.createButton("trash", () => this.cancel())
+    };
+
+    this.controlRow = $("<div class='controls'>").appendTo(this.cropContainer);
+
+    // rotation buttons on the left
+    this.controlRow.append($("<div class='btn-group' role='group'>")
+      .append(this.controls.rotateLeft).append(this.controls.rotateRight));
+
+    // preview images in the middle
+    this.controlRow.append($("<div class='previews'>")
+      .append("<div class='preview-lg'>")
+      .append("<div class='preview-md'>")
+      .append("<div class='preview-sm'>"));
+
+    // main buttons on the right
+    this.controlRow.append($("<div class='btn-group buttons-main' role='group'>")
+      .append(this.controls.reset).append(this.controls.cancel).append(this.controls.accept));
+  }
+
+  /**
+   * Called when the user clicked accept button. Sets file data and triggers file upload.
+   */
+  cropImage() {
+    const canvas = this.cropper.getCroppedCanvas(null);
+
+    // replace the stored file with the new canvas
+    this.fineUploader.clearStoredFiles();
+    this.fineUploader.addFiles([{
+      canvas: canvas,
+      name: this.fileName,
+      quality: 100,
+      type: this.mimeType
+    }]);
+
+    // reset all controls
+    this.cancel();
+    this.picture.setAttribute("src", canvas.toDataURL(this.mimeType));
+
+    // finally start uploading
+    this.setLoading(true);
+    this.fineUploader.uploadStoredFiles();
+  }
+
+  /**
+   * Is called after the file upload has been completed and the profile photo changed.
+   * @param {number} id - The current file's id.
+   * @param {string} fileName - The current file's name.
+   * @param {object} responseJSON - The server's json response.
+   */
+  onUploadCompleted(id, fileName, responseJSON) {
+    this.setLoading(false);
+    this.fileInput.classList.remove("hidden");
+    this.info.innerText = Diaspora.I18n.t("photo_uploader.completed", {"file": fileName});
+
+    if (responseJSON.data !== undefined) {
```

Comment:
What if uploading failed for server-side reasons?

Reply:
In this case the following error is shown in the UI. The UI is changed back to initial state (same behavior as when image cropping process is cancelled by the user).
![upload-error](https://user-images.githubusercontent.com/1684826/34428398-b4700cac-ec4c-11e7-9dc7-124677017751.png)